IdentityCore/src/validation/MSIDCIAMAuthority.m

```diff
@@ -0,0 +1,184 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+#if !EXCLUDE_FROM_MSALCPP
+
+#import "MSIDCIAMAuthority.h"
+#import "MSIDCIAMAuthorityResolver.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDAuthority+Internal.h"
+#import "MSIDJsonSerializableFactory.h"
+#import "MSIDJsonSerializableTypes.h"
+#import "MSIDProviderType.h"
+#import "MSIDAADTenant.h"
+#import "MSIDAADAuthority.h"
+#import "NSString+MSIDExtensions.h"
+
+@implementation MSIDCIAMAuthority
+
++ (void)load
+{
+    [MSIDJsonSerializableFactory registerClass:self forClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+    [MSIDJsonSerializableFactory mapJSONKey:MSID_PROVIDER_TYPE_JSON_KEY keyValue:MSID_JSON_TYPE_PROVIDER_CIAM kindOfClass:MSIDAuthority.class toClassType:MSID_JSON_TYPE_CIAM_AUTHORITY];
+}
+
+- (nullable instancetype)initWithURL:(nonnull NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                           rawTenant:(nullable NSString *)rawTenant
+                             context:(nullable id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{    
+    self = [self initWithURL:url validateFormat:validateFormat context:context error:error];
+    if (self)
+    {
+        if (rawTenant)
+        {
+            if ([self.class isAuthorityFormatValid:url context:context error:nil])
+            {
+                _url = [NSURL URLWithString:[NSString stringWithFormat:@"https://%@/%@", [url msidHostWithPortIfNecessary], rawTenant]];
+                _realm = rawTenant;
+            }
+        }
+    }
+
+    return self;
+}
+
+- (nullable instancetype)initWithURL:(NSURL *)url
+                      validateFormat:(BOOL)validateFormat
+                             context:(id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    self = [super initWithURL:url validateFormat:validateFormat context:context error:error];
+    
+    NSArray *hostComponents = [url.msidHostWithPortIfNecessary componentsSeparatedByString:@"."];
+    
+    //If we have the URL tenant.ciamlogin.com
+    if (url.pathComponents.count == 0 || [[url lastPathComponent] isEqual:@"/"])
```

Comment:
Wouldn't this incorrectly append xxx.ciamlogin.com if we have a slash in the end?
e.g. if input authority is https://login.microsoftonline.com/mytenant/?